src/backend/optimizer/plan/planner.c

```diff
@@ -1866,6 +1907,386 @@ grouping_planner(PlannerInfo *root, double tuple_fraction)
 		 */
 		current_rel = query_planner(root, standard_qp_callback, &qp_extra);
 
+		/*
+		 * Answer Query Using Materialized Views(AQUMV).
+		 */
+		RelOptInfo *mv_final_rel = NULL;
+		bool can_not_use_mv = (parse->commandType != CMD_SELECT) ||
+							  (parse->rowMarks != NIL) ||
+							  parse->hasAggs ||
+							  parse->hasWindowFuncs ||
+							  parse->hasDistinctOn ||
+							  parse->hasModifyingCTE ||
+							  parse->sortClause ||
+							  (parse->parentStmtType == PARENTSTMTTYPE_REFRESH_MATVIEW) ||
+							  (parse->parentStmtType == PARENTSTMTTYPE_CTAS) ||
+							  parse->hasSubLinks;
+
+		if (Gp_role == GP_ROLE_DISPATCH &&
+			answer_query_using_materialized_views &&
+			!can_not_use_mv &&
+			(list_length(parse->jointree->fromlist) == 1))
+		{
+			Node	   *jtnode = (Node *) linitial(parse->jointree->fromlist);
+
+			if (IsA(jtnode, RangeTblRef))
+			{
+				int			varno = ((RangeTblRef *) jtnode)->rtindex;
+				RangeTblEntry *rte = root->simple_rte_array[varno];
+
+				Assert(rte != NULL);
+
+				/*
+				 * AQUMV_FIXME_MVP:
+				 *	Single relation, excluding catalog.
+				 */
+				if (rte->rtekind == RTE_RELATION &&
+					!IsSystemClassByRelid(rte->relid) &&
+					!has_superclass(rte->relid) &&
+					!has_subclass(rte->relid))
+				{
+
+					Relation		ruleDesc;
+					Relation		matviewRel;
+					SysScanDesc		rcscan;
+					HeapTuple		tup;
+					RewriteRule		*rule;
+					Form_pg_rewrite	rewrite_tup;
+					List			*actions;
+					Query			*mvQuery;
+					Query			*mvRelQueryTree; /* Query tree of select from mv itself. */
+					Node			*mvjtnode;
+					bool			need_close = false;
+
+					ruleDesc = table_open(RewriteRelationId, AccessShareLock);
+
+					rcscan = systable_beginscan(ruleDesc, InvalidOid, false,
+												NULL, 0, NULL);
+
+					while (HeapTupleIsValid(tup = systable_getnext(rcscan)))
+					{
+						CHECK_FOR_INTERRUPTS();
+						if (need_close)
+							table_close(matviewRel, AccessShareLock);
+
+						rewrite_tup = (Form_pg_rewrite) GETSTRUCT(tup);
+
+						matviewRel = table_open(rewrite_tup->ev_class, AccessShareLock);
+						need_close = true;
+
+						/*
+						 * AQUMV
+						 * Currently the data of IVM is always up-to-date.
+						 * Take care of this when IVM defered-fefresh is supported.
+						 */
+						if (!RelationIsIVM(matviewRel))
```

Comment:
need consider `RelationIsPopulated()` , in case of `WITH NO DATA ` ?

Reply:
Good catch.

Reply:
Fixed, a case is added to test mv with no data.

Reply:
great!